core/deliverservice/bft_header_receiver.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package deliverclient
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/hyperledger/fabric/protos/common"
+	"github.com/hyperledger/fabric/protos/orderer"
+)
+
+const bftHeaderWrongStatusThreshold = 10
+
+type bftHeaderReceiver struct {
+	mutex             sync.Mutex
+	chainID           string
+	minBackoffDelay   time.Duration
+	maxBackoffDelay   time.Duration
+	stop              bool
+	stopChan          chan struct{}
+	started           bool
+	endpoint          string
+	client            *broadcastClient
+	msgCryptoVerifier MessageCryptoVerifier
+	lastHeader        *common.Block // a block with Header & Metadata, without Data (i.e. lastHeader.Data==nil)
+	lastHeaderTime    time.Time
+}
+
+func newBFTHeaderReceiver(
+	chainID string,
+	endpoint string,
+	client *broadcastClient,
+	msgVerifier MessageCryptoVerifier,
+	minBackOff time.Duration,
+	maxBackOff time.Duration,
+) *bftHeaderReceiver {
+	hRcv := &bftHeaderReceiver{
+		chainID:           chainID,
+		stopChan:          make(chan struct{}, 1),
+		endpoint:          endpoint,
+		client:            client,
+		msgCryptoVerifier: msgVerifier,
+		minBackoffDelay:   minBackOff,
+		maxBackoffDelay:   maxBackOff,
+	}
+	return hRcv
+}
+
+func (hr *bftHeaderReceiver) DeliverHeaders() {
+	defer func() {
+		hr.Close()
+	}()
+
+	bftLogger.Debugf("[%s] Starting to deliver headers from endpoint: %s", hr.chainID, hr.endpoint)
+	hr.setStarted()
+	var errorStatusCounter int
+	var statusCounter uint
+
+	for !hr.isStopped() {
+		msg, err := hr.client.Recv()
+		if err != nil {
+			bftLogger.Warningf("[%s] Receive error: %s", hr.chainID, err.Error())
+			return
+		}
+
+		switch t := msg.Type.(type) {
+		case *orderer.DeliverResponse_Status:
+			if t.Status == common.Status_SUCCESS {
+				bftLogger.Warningf("[%s] ERROR! Received success for a seek that should never complete", hr.chainID)
+				return
+			}
+			if t.Status == common.Status_BAD_REQUEST || t.Status == common.Status_FORBIDDEN {
+				bftLogger.Errorf("[%s] Got error %v", hr.chainID, t)
+				errorStatusCounter++
+				if errorStatusCounter > bftHeaderWrongStatusThreshold {
+					bftLogger.Criticalf("[%s] Wrong statuses threshold passed, stopping bft header receiver", hr.chainID)
+					return
+				}
+			} else {
+				errorStatusCounter = 0
+				bftLogger.Warningf("[%s] Got error %v", hr.chainID, t)
+			}
+			dur := backOffDuration(2.0, statusCounter, hr.minBackoffDelay, hr.maxBackoffDelay)
+			bftLogger.Debugf("[%s] going to retry in: %s", hr.chainID, dur)
+			backOffSleep(dur, hr.stopChan)
+			statusCounter++
+
+			hr.client.Disconnect()
+			continue
+
+		case *orderer.DeliverResponse_Block:
```

Comment:
we should verify when we get the block, not afterwards.

Reply:
we verify the blocks when we ask for LastBlockNum(), in order to avoid verifying blocks that are overwritten and never read.